src/net_processing.cpp

```diff
@@ -3823,6 +3829,10 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
                     }
 
                     pto->setInventoryTxToSend.insert(setRebroadcastTxs.begin(), setRebroadcastTxs.end());
+
+                    // also ensure inclusion of wallet txns that haven't been successfully broadcast yet
+                    // since set elements are unique, this will be a no-op if the txns are already in setInventoryTxToSend
+                    pto->setInventoryTxToSend.insert(mempool.setUnbroadcastTxIDs.begin(), mempool.setUnbroadcastTxIDs.end());
```

Comment:
If we add wallet transactions to the rebroadcast INV that have a smaller feerate than the top of our mempool, wouldn’t there be a feerate gap to the other INVs of the message, making the wallet transactions easily identifiable as such and reducing privacy?

Reply:
> making the wallet transactions easily identifiable as such and reducing privacy?

I don't think this is true (at least no worse than today). While it is possible to guess whether a list of INVs is a rebroadcast inv or not, it shouldn't be possible to trivially find the source of the low feerate txs in that inv. Txs broadcast for the first time from our node should be no different from txs broadcast for the first time from another node to us and then relayed by us. It is know to not be perfect (see dandelion tx relay), but improving that seems like a separate issue to me.

Reply:
Want to clarify my expectations of the behavior- 

We add transactions to `setInventoryTxToSend` when we expect to relay the transaction to a peer. This includes if we are the source node for the transaction as well as if we are relaying it. With these changes, both the transactions to rebroadcast and the "unbroadcast" transactions are added to `setInventoryTxToSend` on the rebroadcast poisson timer. 

Later, all transactions from `setInventoryTxToSend` get added to `vInvTx`, passed through some checks, then relayed to peers. 

So, what this means... 
* if they are keeping track, peers can detect rebroadcast transactions because `INV` messages have been sent for that transaction before. 
* `INVs` for unbroadcast transactions have probably not been sent out before (unless all the peers who received the message did not send a `GETDATA` back). 
* In the normal case, the peer would not be able to distinguish a transaction sent from the unbroadcast set from a transaction being submitted initially, or the initial relay. This is known to be imperfect, but these changes should not make privacy worse than current behavior. 
* The only new potential for privacy leak that I see based on these changes- 
A node disables p2p, creates a transaction, enables p2p. At that time, that transaction gets propagated out to the network due to being on the unbroadcast set. The first peer the node sends an `INV` to is the attacker. The attacker is able to send back a `GETDATA` before any other `INV` messages are sent out. The attacker does not relay the transaction to any other nodes.
Time passes, the node calculates the transaction "should have been mined" by now, puts it into the rebroadcast set, relays it out to its peers. The attacker has another connection open to the node & is able to identify that it is a rebroadcast. The attacker would also have to have many other nodes & connections to the rest of the network to identify that this is the only node rebroadcasting that transaction. This seems pretty difficult to execute, definitely harder to discern source node based on rebroadcast logic than in the existing implementation. 
* In this case, the much bigger concern would be the ability of an attacker to stop the propagation of a transaction by sending the `GETDATA` back very quickly, thus removing it from the node's unbroadcast set. I'm trying to better understand the risk of this & brainstorm possible better solutions (see https://github.com/bitcoin/bitcoin/pull/16698#discussion_r348840194). If any reviewers have thoughts or questions about this mechanism, let's continue the convo on that thread. 

@mzumsande to conclude this long explanation, hopefully this resolves your original question? The unbroadcast set should have (pretty much) the same guarantees as initial relay. LMK if you have any outstanding questions or recommendations for improved documentation.